src/cli.rs

```diff
@@ -1,22 +1,120 @@
-use crate::R;
+use clap::{App, Arg, Error};
 use std::path::PathBuf;
 
+#[derive(Debug, PartialEq)]
 pub enum Args {
     ExecutableMock { executable_mock_path: PathBuf },
-    CheckProtocols { script_path: PathBuf },
+    CheckProtocols { script_path: PathBuf, record: bool },
 }
 
-pub fn parse_args(mut args: impl Iterator<Item = String>) -> R<Args> {
-    args.next()
-        .ok_or("argv: expected program name as argument 0")?;
-    Ok(match args.next().ok_or("supply one argument")?.as_ref() {
-        "--executable-mock" => Args::ExecutableMock {
+pub fn parse_args(args: impl Iterator<Item = String>) -> Args {
+    match parse_args_safe(args) {
+        Ok(args) => args,
+        Err(error) => error.exit(),
+    }
+}
+
+fn parse_args_safe(args: impl Iterator<Item = String>) -> Result<Args, Error> {
+    let args: Vec<_> = args.collect();
+    if args.get(1) == Some(&"--executable-mock".to_string()) {
+        Ok(Args::ExecutableMock {
             executable_mock_path: PathBuf::from(
-                args.next().expect("expected executable file as argument 1"),
+                args.get(2).expect("missing argument: executable mock file"),
             ),
-        },
-        argument => Args::CheckProtocols {
-            script_path: PathBuf::from(argument),
-        },
-    })
+        })
+    } else {
+        let matches = App::new("check-protocols")
+            .arg(Arg::with_name("record").short("r").long("record").help(
+                "[EXPERIMENTAL] Runs the script (without mocking out anything), \
+                 records a protocol and prints it to stdout",
+            ))
+            .arg(
+                Arg::with_name("program")
+                    .help("the program to test")
+                    .required(true)
+                    .index(1),
+            )
+            .get_matches_from_safe(args)?;
+        Ok(Args::CheckProtocols {
+            script_path: PathBuf::from(matches.value_of("program").unwrap()),
```

Comment:
it's a force of habit of mine from scala, and i'm not sure if it fully carries over to rust, but even though the arg is required and it should be safe, i would still consider using `ok_or` over `unwrap`.